docs/examples/creating-a-crud-api.md

```diff
@@ -0,0 +1,269 @@
+---
+description: |
+  Use HTTP CRUD methods to perform operations on resources. Learn how to use HTTP handlers to create a RESTful API.
+---
+
+The MDN [docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods) are a
+great resource to learn more about HTTP methods. We'll touch on the four
+fundamental methods necessary to create a basic CRUD (create, read, update,
+delete) API here. Additionally, we'll briefly mention CORS requests and how
+`OPTIONS` comes into play.
+
+Using HTTP methods is a common way to create a REST API. Fresh supports common
+HTTP methods in handlers out of the box. Async HTTP requests are also supported.
+Read more about custom handlers [here](/docs/getting-started/custom-handlers).
+
+In this example we'll be creating a small API that uses
+[Deno KV](https://deno.com/kv) to store users in a database.
+
+Our project structure will look like this (in addition to the rest of the Fresh
+code from a new project):
+
+```
+├── routes
+│   └── api
+│       └── users
+│           ├── [id].ts
+│           └── index.ts
+```
+
+In each section about a method, only the relevant handler will be shown. The
+full files are available at the bottom for reference.
+
+## POST
+
+`POST` (create) is used to create a resource.
+
+```tsx
+// /routes/api/index.ts
+async POST(req, _ctx) {
+  const body = await req.text();
+  const user = JSON.parse(body) as User;
+  await upsertUser(user);
+  return new Response(JSON.stringify(user));
+},
+```
+
+Test this with Postman (or your favorite client) with a URL like
+`http://localhost:8000/api/users` and a method of `POST`. Make sure to have a
+payload like:
+
+```json
+{
+  "id": "2",
+  "name": "TestUserName"
+}
+```
+
+You should receive the same thing back:
+
+```json
+{ "id": "2", "name": "TestUserName" }
+```
+
+## GET
+
+`GET` (read) is used to retrieve a resource and is by far the most common HTTP
+method. You can use `GET` to fetch database content, markdown, or static files.
+
+```tsx
+// /routes/api/[id].ts
+async GET(_req, ctx) {
+  const id = ctx.params.id;
+  const user = await getUserById(id);
+  return new Response(JSON.stringify(user));
+},
+```
+
+Let's practice retrieving our user! A `GET` request to
+`http://localhost:8000/api/users/2` should return:
+
+```json
+{ "id": "2", "name": "TestUserName" }
+```
+
+## PUT (and PATCH)
+
+`PUT` (update) and `PATCH` are used to update a resource. While similar, there
+are differences and you should use the one that best suits your use case. Read
+more about HTTP methods on
+[MDN](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods).
+
+An example of an update endpoint using `PUT`:
+
+```tsx
+// /routes/api/[id].ts
+async PUT(req, ctx) {
+  const id = ctx.params.id;
+  const body = await req.text();
+  const userBody = JSON.parse(body) as User;
+  updateUserById(id, userBody);
+  return new Response(JSON.stringify(userBody));
+},
+```
+
+Time to change their name. We'll now `PUT` a request to
```

Comment:
Might be me not being a native speaker, but I got confused for a second to what "their" is referring to here.

```suggestion
Time to change a user's name. We'll now `PUT` a request to
```

Reply:
I'll push back 💪  here as a native speaker. We're still talking about user 2, but we don't know whether user 2 is male or female (or if they identify differently, and it's entirely irrelevant as well), so we can just use "their". See https://en.wikipedia.org/wiki/Singular_they for details. (I was just talking about this on Sunday with a friend from Berlin, so you're not alone!)

If you still find it confusing, I would go for something like `Time to change their name from "TestUserName" to something else` or `Time to change user 2's name`.

Reply:
Right that makes sense. Thanks for the explanation 👍